### ISSUE TITLE
With react-router, redux will not work.

### ISSUE THREAD

**Issue by ustccjw** [gaearon/redux#136]
I'm a flummox's user, as @acdlite suggest, i try to use redux.

I use react-router v1.0-beta2, I found when I use {this.props.children} to use 'component of redux', @connect, then will report error: `Failed Context Types: Required context 'redux' was not specified in 'Connector'. Check the render method of 'Connector(DashboardApp)'`

It seems react-router disabled the 'redux' context, so what should I do ?

**Comment by ooflorent** [comment]
Can you attach a gist?

**Comment by ustccjw** [comment]
@provide(redux)
    export default class App extends React.Component {
    	constructor(props, context) {
    		super(props)
    	}
    
    	static childContextTypes = {
    		muiTheme: React.PropTypes.object
    	}
    
    	getChildContext() {
    		return {
    			muiTheme: ThemeManager.getCurrentTheme()
    		}
    	}
    
    	render() {
    		return (
    			<div className="App">
    				<Header />
    				{this.props.children}
    			</div>
    		)
    	}
    }


@connect(state => ({
    	counter: state.dashboard
    }))
    export default class DashboardApp extends React.Component {
    	constructor(props, context) {
    		super(props)
    	}
    
    	render() {
    		const {counter, dispatch} = this.props;
    		return (
    			<Dashboard counter={counter}
    				{...bindActionCreators(DashboardActions, dispatch)} />
    		)
    	}
    }

**Comment by ooflorent** [comment]
Including the router :wink:

**Comment by ustccjw** [comment]
const rootRoute = {
    	path: '/',
    	childRoutes: [
    		dashboard
    	],
    	component: App
    }
    
    const dashboard = {
    	path: 'dashboard',
    	getComponents(cb) {
    		require.ensure([], (require) => {
    			cb(null, require('../containers/dashboard-app'))
    		})
    	}
    }

**Comment by ooflorent** [comment]
You must pass the `context` to `super`.

**Comment by ustccjw** [comment]
@ooflorent i update super(props, context), i always report the same error.

**Comment by gaearon** [comment]
It won't work this way with Router 1.0, until React 0.14 comes out.

Top-level route handler (`App`) is not an owner of the lower level handlers (`DashboardApp`) so the context does not propagate. This will be fixed in React 0.14.

In the meantime, you need to wrap `<Router>` into `<Provider>` (I don't see where you render `<Router>` so can't give an exact example.)

**Comment by emmenko** [comment]
@ustccjw this should help you to understand

https://github.com/emmenko/redux-react-router-async-example/commit/c8e886f16061ae191e02a48f6393a1b700531326

The places to pay attention are:
- how `Provider` wraps the routes
- how to use `@connect` and pass `actions` (or extra `props`) to the child route

```js
{React.cloneElement(this.props.children, { actions })}
```

**Comment by gaearon** [comment]
Let's create a dedicated FAQ in the docs (separate from README) and put RR integration instructions there. Anybody want to do that?

**Comment by ustccjw** [comment]
@gaearon I can do it, i will try my best.

**Comment by ustccjw** [comment]
@emmenko thanks, i learn a lot.

**Comment by emmenko** [comment]
@gaearon I can do a PR later

**Comment by emmenko** [comment]
@ustccjw glad to help :)

**Comment by ustccjw** [comment]
@emmenko in this way, can not hot-load store https://github.com/gaearon/redux/issues/117

**Comment by emmenko** [comment]
@ustccjw to be fair, I haven't tried hot reloading in my example. What exactly is not working? Is it working without RR?

**Comment by ustccjw** [comment]
@emmenko #117 `import your stores in App.js (root component), not index.js`, so  if import stores in index.js, the hot-loading will be failed.

As @gaearon said, this will wait react V0.14, then can import stores in App

    import * as stores from './stores'
    ...
    @provide(redux)
    export default class App {
    }

**Comment by gaearon** [comment]
To have hot reloading properly working, you'll need to have a root component that returns `<Router>` wrapped in `<Provider>` from its `render`, and imports stores.

If you import stores in `index.js`, hot reloading them indeed won't work.

**Comment by ustccjw** [comment]
@gaearon you are right, it works! But react-router will throw error: Uncaught Error: Invariant Violation: <Router history> may not be changed.

**Comment by gaearon** [comment]
@ustccjw 

Thanks for reporting!
You should be able to do something like

```js
class AppContainer extends Component {
  constructor(props) {
    this.history = new BrowserHistory();
    super(props);
  }

  render() {
    return (
      <Provider>{() =>
        <Router history={this.history}>
           ...
        </Router>
      </Provider>
    );
  }
```

@mjackson @ryanflorence No plans to support swapping `history` object on the go?

**Comment by ustccjw** [comment]
Thanks! You are great, I will to learn more from you！

ps. this can not use before super(), this should be

    constructor(props) {
        super(props)
        this.history = new BrowserHistory()
    }

**Comment by gaearon** [comment]
Right.

**Comment by gaearon** [comment]
@emmenko Can you please fix your RR project so hot reloading works taking this comment into account? https://github.com/gaearon/redux/issues/136#issuecomment-113498784

**Comment by emmenko** [comment]
@gaearon sure, I wanted to make it work first without HR, I'll do that as a second step

**Comment by justjacksonn** [comment]
Not sure if this is related... and/or I am stuck waiting for React final 0.14.. I am currently using 0.14 rc1. I would like to render a top header, bottom footer and then have my router paths render in between the header and footer areas.. the content area. However, my header is a top menu bar, with <Link...> elements in it.. and every time I use those, I get some sort of infinite loop happening and my entire browser halts.. I have to force quit it.

I assume this has something to do with that the header and footer components are rendered outside of the Provider? Is there some way to make the Link element (or even an <a> element) work to change the content area of my page to the specified routed path/component? I would be fine with header/footer as well as content area inside the Provider tag.. but when I do that, the entire page gets redrawn including header/footer every time I click a link, and I am still stuck with the <Link> not working and crashing browser in the header component.

In other words, I am still not quite clear on the proper setup/integration of components with Redux and router. :)  Appreciate any help on this.

**Comment by gaearon** [comment]
Sorry, it is very hard what goes wrong in your particular case but I assure you it is not combining Redux and React Router itself that is the problem, but rather a mistake in using either of them (or something else entirely).

Please consult `examples/real-world` for React Router integration. If you still have this issue please provide a sample project reproducing it.

**Comment by adriennetran** [comment]
@gaearon Thanks! I tried your code and now I get a different error `ReferenceError: Link is not defined`.
